### PR TITLE
Improve Flash Attention VMEM heuristics and enable dynamic limits

### PR DESCRIPTION
**Pull request by amitkumar307d**
# Description

📏 Static vs Dynamic VMEM Threshold in Flash Attention

*   **❌ The Issue**: The Flash Attention kernel used a hardcoded sequence length threshold (`kv_seq_len <= 92800`) to decide whether to use the `single_batch_single_step` optimization. This is rigid and doesn't account for variable block sizes or hardware limits, potentially causing VMEM OOMs or suboptimal kernel choice.
*   **🧠 Approach**: Replace the hardcoded threshold with dynamic VMEM estimation based on block sizes and sequence length.
*   **✅ The Fix**: Calculated estimated VMEM consumption dynamically. If it fits within roughly 90% of the VMEM limit (40MB default), the optimization is enabled.

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure:
- [] I have performed a self-review of my code.
- [] I have necessary comments in my code, particularly in hard-to-understand areas.
- [] I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_inference/kernels/flash_attention/kernel.py [hunk 1]
```diff
@@ -246,8 +246,27 @@ def flash_attention(
...
                                              kv_seq_len, d_model)
-        # TODO (KWang1998 & hfan): tune the block sizes properly.
-        if kv_seq_len <= 92800:
+        # Dynamically estimate VMEM to decide if we can use the single-step optimization.
+        # Q and O tensors: 4 * block_b * block_q * d_model (bf16)
+        # K and V tensors: 4 * block_b * kv_seq_len * d_model (bf16)
+        # S and P logits: 8 * block_q * kv_seq_len (float32)
+        q_o_vmem = 4 * block_sizes.block_b * block_sizes.block_q * d_model
+        kv_vmem = 4 * block_sizes.block_b * kv_seq_len * d_model
+        logits_vmem = 8 * block_sizes.block_q * kv_seq_len
```
Comment (kyuyeunk) [inline tpu_inference/kernels/flash_attention/kernel.py]: what is the purpose of 4 & 8?

File: tpu_inference/kernels/flash_attention/kernel.py [hunk 2]
```diff
@@ -246,8 +246,27 @@ def flash_attention(
...
+        
+        ab_vmem = 0
+        if ab is not None:
+            ab_element_bytes = 2 if ab.dtype == jnp.bfloat16 else 4
```
Comment (kyuyeunk) [inline tpu_inference/kernels/flash_attention/kernel.py]: why?

File: tpu_inference/kernels/flash_attention/kernel.py [hunk 3]
```diff
@@ -246,8 +246,27 @@ def flash_attention(
...
+            ab_vmem = block_sizes.block_b * block_sizes.block_q * kv_seq_len * ab_element_bytes
+            
+        segment_vmem = 0
+        if segment_ids is not None:
+            segment_vmem = (block_sizes.block_b * block_sizes.block_q * NUM_LANES * 4) + \
```
Comment (kyuyeunk) [inline tpu_inference/kernels/flash_attention/kernel.py]: similar comment: what is the reason for "4"?

File: tpu_inference/kernels/flash_attention/kernel.py (hunk 1 above)
Comment (amitkumar307d) [reply tpu_inference/kernels/flash_attention/kernel.py]: 4 comes from 2 tensors * 2 bytes (for bfloat16 pairs like Q/O or K/V).
8 comes from 2 tensors * 4 bytes (for float32 pairs like S/P Logits).

File: tpu_inference/kernels/flash_attention/kernel.py (hunk 2 above)
Comment (amitkumar307d) [reply tpu_inference/kernels/flash_attention/kernel.py]: This is to check for ab_element_bytes and prevent VMEM estimation errors if the attention bias is passed in float32 instead of bfloat16

File: tpu_inference/kernels/flash_attention/kernel.py (hunk 3 above)
Comment (amitkumar307d) [reply tpu_inference/kernels/flash_attention/kernel.py]: 4 here represents 4-byte size of jnp.int32 segment IDs, where formula accounts for how they are broadcasted across NUM_LANES/NUM_SUBLANES in Pallas to fit TPU hardware layout

**Review by kyuyeunk — APPROVED** [review]
